File: backend/app/services/analytics_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func, and_
from app.models.survey import SurveyResponse
from app.models.reminder_log import ReminderLog, StudentLocation
from app.models.user import User
from app.models.course import Course, CourseEnrollment
from app.models.timetable import TimetableEntry
from app.core.redis import get_location
from app.utils.week_helpers import get_week_for_date
from datetime import date, datetime, timezone, timedelta
import pandas as pd
import numpy as np


async def get_punctuality_summary(db: AsyncSession):
    result = await db.execute(select(SurveyResponse))
    rows = result.scalars().all()
    if not rows:
        return {"avg_punctuality": None, "avg_missed": None, "total_responses": 0}
    df = pd.DataFrame([{
        "week": r.survey_week,
        "q1": r.q1_punctuality_rating,
        "q2": r.q2_missed_classes,
        "q3": r.q3_reminder_helpful,
        "q5": r.q5_privacy_comfort,
    } for r in rows])
    return {
        "avg_punctuality": round(float(df["q1"].mean()), 2),
        "avg_missed": round(float(df["q2"].mean()), 2),
        "total_responses": len(df),
        "helpful_pct": round(float(df["q3"].mean() * 100), 1),
        "avg_privacy_comfort": round(float(df["q5"].mean()), 2),
    }


async def get_weekly_trend(db: AsyncSession):
    result = await db.execute(select(SurveyResponse).order_by(SurveyResponse.survey_week))
    rows = result.scalars().all()
    if not rows:
        return []
    df = pd.DataFrame([{
        "week": r.survey_week,
        "q1": r.q1_punctuality_rating,
        "q2": r.q2_missed_classes,
    } for r in rows])
    grouped = df.groupby("week").agg(
        avg_punctuality=("q1", "mean"),
        avg_missed=("q2", "mean"),
        response_count=("q1", "count"),
    ).reset_index()

    rl = await db.execute(select(ReminderLog))
    rl_rows = rl.scalars().all()
    reminder_counts = {}
    if rl_rows:
        rdf = pd.DataFrame([{"class_date": r.class_date} for r in rl_rows])
        rdf["week"] = rdf["class_date"].apply(get_week_for_date)
        reminder_counts = rdf.groupby("week").size().to_dict()

    trend = []
    for _, row in grouped.iterrows():
        w = int(row["week"])
        trend.append({
            "week": w,
            "avg_punctuality": round(float(row["avg_punctuality"]), 2),
            "avg_missed": round(float(row["avg_missed"]), 2),
            "response_count": int(row["response_count"]),
            "reminder_count": reminder_counts.get(w, 0),
        })
    return trend


async def get_reminder_effectiveness(db: AsyncSession):
    surveys = await db.execute(select(SurveyResponse))
    s_rows = surveys.scalars().all()
    reminders = await db.execute(
        select(ReminderLog.student_id, func.count(ReminderLog.id).label("cnt"))
        .group_by(ReminderLog.student_id)
    )
    r_rows = reminders.all()

    if not s_rows or not r_rows:
        return {"correlation": None, "interpretation": "Insufficient data"}

    s_map = {}
    for s in s_rows:
        sid = str(s.student_id)
        if sid not in s_map:
            s_map[sid] = []
        s_map[sid].append(s.q1_punctuality_rating)

    r_map = {str(r.student_id): r.cnt for r in r_rows}
    data = []
    for sid, ratings in s_map.items():
        if sid in r_map:
            data.append({"reminder_count": r_map[sid], "avg_q1": np.mean(ratings)})

    if len(data) < 3:
        return {"correlation": None, "interpretation": "Insufficient data"}

    df = pd.DataFrame(data)
    corr = float(df["reminder_count"].corr(df["avg_q1"]))

    if corr > 0.5:
        interpretation = "Strong positive correlation — more reminders associated with higher punctuality ratings"
    elif corr > 0.2:
        interpretation = "Moderate positive correlation — students who received more reminders rated punctuality slightly higher"
    elif corr > -0.2:
        interpretation = "Weak or no correlation — reminder count shows little relationship with punctuality ratings"
    else:
        interpretation = "Negative correlation — unexpected pattern; review data quality"

    return {"correlation": round(corr, 3), "interpretation": interpretation}


async def get_platform_breakdown(db: AsyncSession):
    result = await db.execute(
        select(User.platform, func.count(User.id))
        .where(User.role == "student", User.is_active == True)
        .group_by(User.platform)
    )
    rows = result.all()
    breakdown = {"android": 0, "ios": 0, "unknown": 0}
    for platform, count in rows:
        key = str(platform.value) if platform else "unknown"
        breakdown[key] = count

    rl = await db.execute(
        select(ReminderLog.student_id)
    )
    return breakdown


async def get_geofence_stats(db: AsyncSession):
    today = date.today()
    result = await db.execute(
        select(ReminderLog)
        .where(ReminderLog.class_date == today)
    )
    today_logs = result.scalars().all()
    on_campus = sum(1 for l in today_logs if l.was_on_campus)
    off_campus = sum(1 for l in today_logs if not l.was_on_campus)
    return {
        "reminders_today": len(today_logs),
        "on_campus_reminders": on_campus,
        "off_campus_reminders": off_campus,
    }

File: backend-php/src/Services/AnalyticsService.php
<?php
declare(strict_types=1);

namespace App\Services;

use App\Core\Database;
use App\Core\Config;

class AnalyticsService
{
    public static function getPunctualitySummary(): array
    {
        $pdo  = Database::getConnection();
        $rows = $pdo->query('SELECT q1_punctuality_rating, q2_missed_classes, q3_reminder_helpful, q5_privacy_comfort FROM survey_responses')->fetchAll();

        if (empty($rows)) {
            return ['avg_punctuality' => null, 'avg_missed' => null, 'total_responses' => 0];
        }

        $q1 = array_column($rows, 'q1_punctuality_rating');
        $q2 = array_column($rows, 'q2_missed_classes');
        $q3 = array_column($rows, 'q3_reminder_helpful');
        $q5 = array_column($rows, 'q5_privacy_comfort');

        return [
            'avg_punctuality'    => round(array_sum($q1) / count($q1), 2),
            'avg_missed'         => round(array_sum($q2) / count($q2), 2),
            'total_responses'    => count($rows),
            'helpful_pct'        => round((array_sum(array_map(fn($v) => (int)$v, $q3)) / count($q3)) * 100, 1),
            'avg_privacy_comfort' => round(array_sum($q5) / count($q5), 2),
        ];
    }

    public static function getWeeklyTrend(): array
    {
        $pdo  = Database::getConnection();
        $rows = $pdo->query('SELECT survey_week, q1_punctuality_rating, q2_missed_classes FROM survey_responses ORDER BY survey_week')->fetchAll();

        if (empty($rows)) {
            return [];
        }

        $byWeek = [];
        foreach ($rows as $row) {
            $w = (int)$row['survey_week'];
            if (!isset($byWeek[$w])) {
                $byWeek[$w] = ['q1' => [], 'q2' => [], 'count' => 0];
            }
            $byWeek[$w]['q1'][]   = (float)$row['q1_punctuality_rating'];
            $byWeek[$w]['q2'][]   = (float)$row['q2_missed_classes'];
            $byWeek[$w]['count']++;
        }

        $studyStart = new \DateTime(Config::getInstance()->get('study_start_date', '2024-09-01'));
        $reminders  = $pdo->query('SELECT class_date FROM reminder_logs')->fetchAll();
        $reminderCounts = [];
        foreach ($reminders as $r) {
            $d    = new \DateTime($r['class_date']);
            $diff = $studyStart->diff($d)->days;
            $week = max(1, (int)($diff / 7) + 1);
            $reminderCounts[$week] = ($reminderCounts[$week] ?? 0) + 1;
        }

        $trend = [];
        foreach ($byWeek as $week => $data) {
            $trend[] = [
                'week'            => $week,
                'avg_punctuality' => round(array_sum($data['q1']) / count($data['q1']), 2),
                'avg_missed'      => round(array_sum($data['q2']) / count($data['q2']), 2),
                'response_count'  => $data['count'],
                'reminder_count'  => $reminderCounts[$week] ?? 0,
            ];
        }
        return $trend;
    }

    public static function getReminderEffectiveness(): array
    {
        $pdo     = Database::getConnection();
        $surveys = $pdo->query('SELECT student_id, q1_punctuality_rating FROM survey_responses')->fetchAll();
        $reminders = $pdo->query('SELECT student_id, COUNT(*) as cnt FROM reminder_logs GROUP BY student_id')->fetchAll();

        if (empty($surveys) || empty($reminders)) {
            return ['correlation' => null, 'interpretation' => 'Insufficient data'];
        }

        $sMap = [];
        foreach ($surveys as $s) {
            $sid = $s['student_id'];
            $sMap[$sid][] = (float)$s['q1_punctuality_rating'];
        }

        $rMap = [];
        foreach ($reminders as $r) {
            $rMap[$r['student_id']] = (int)$r['cnt'];
        }

        $data = [];
        foreach ($sMap as $sid => $ratings) {
            if (isset($rMap[$sid])) {
                $data[] = ['x' => $rMap[$sid], 'y' => array_sum($ratings) / count($ratings)];
            }
        }

        if (count($data) < 3) {
            return ['correlation' => null, 'interpretation' => 'Insufficient data'];
        }

        $n    = count($data);
        $xs   = array_column($data, 'x');
        $ys   = array_column($data, 'y');
        $xMean = array_sum($xs) / $n;
        $yMean = array_sum($ys) / $n;

        $num = 0;
        $dx2 = 0;
        $dy2 = 0;
        for ($i = 0; $i < $n; $i++) {
            $dx  = $xs[$i] - $xMean;
            $dy  = $ys[$i] - $yMean;
            $num += $dx * $dy;
            $dx2 += $dx * $dx;
            $dy2 += $dy * $dy;
        }
        $corr = ($dx2 * $dy2 > 0) ? $num / sqrt($dx2 * $dy2) : 0;

        if ($corr > 0.5) {
            $interpretation = 'Strong positive correlation — more reminders associated with higher punctuality ratings';
        } elseif ($corr > 0.2) {
            $interpretation = 'Moderate positive correlation — students who received more reminders rated punctuality slightly higher';
        } elseif ($corr > -0.2) {
            $interpretation = 'Weak or no correlation — reminder count shows little relationship with punctuality ratings';
        } else {
            $interpretation = 'Negative correlation — unexpected pattern; review data quality';
        }

        return ['correlation' => round($corr, 3), 'interpretation' => $interpretation];
    }

    public static function getPlatformBreakdown(): array
    {
        $pdo  = Database::getConnection();
        $stmt = $pdo->query("
            SELECT platform, COUNT(*) as cnt
            FROM users
            WHERE role = 'student' AND is_active = TRUE
            GROUP BY platform
        ");
        $breakdown = ['android' => 0, 'ios' => 0, 'unknown' => 0];
        foreach ($stmt->fetchAll() as $row) {
            $key = $row['platform'] ?? 'unknown';
            $breakdown[$key] = (int)$row['cnt'];
        }
        return $breakdown;
    }

    public static function getGeofenceStats(): array
    {
        $pdo   = Database::getConnection();
        $today = date('Y-m-d');
        $stmt  = $pdo->prepare('SELECT was_on_campus FROM reminder_logs WHERE class_date = ?');
        $stmt->execute([$today]);
        $logs = $stmt->fetchAll();

        $onCampus  = count(array_filter($logs, fn($l) => (bool)$l['was_on_campus']));
        $offCampus = count($logs) - $onCampus;

        return [
            'reminders_today'     => count($logs),
            'on_campus_reminders' => $onCampus,
            'off_campus_reminders' => $offCampus,
        ];
    }
}
